brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+# Functional Operations
+
+The SOMA API includes functional capabilities around the [SOMA data model](#datamodel.md). The specifics of how these operations manifest in any given language and computing environment is defined in per-language specifications (**to be created**), but each implementation will minimally support the functional operations defined here. For example, it is likely that a Python implementation will prefer `__getattr__` over an explicit `get()` method, and will add other Pythonic functionality to facilitate use.
+
+As any given storage "engine" upon which SOMA is implemented may have additional features and capabilities required to support advanced use cases, it is expected that SOMA implementations will expose engine-specific features. Where possible, these should be implemented to avoid conflict with future changes to the common SOMA API.
+
+> ⚠️ **Issue** - this spec needs to provide guidance on HOW to do that. Fro example, can we carve out some namespace for extensions in the functional API?
+
+In the following, a pseudo-object-oriented syntax is used, eg, `obj.op()` is calling `op()` upon `obj`. Where a type is used, eg, SOMA.op(), it is a `static` operation (ie, no state other than parameters provided).
+
+> ℹ️ **Note** - this section is just a sketch. Primarily focused on adding pseudo-signatures for any primitive operation that must exist on each type.
+
+## SOMAMapping interface
+
+> ℹ️ **Note** - this is an interface defined only to make the subsequent prose simpler. It is not a primary element in the data model.
+
+The SOMAMapping is a string-keyed map. It may be immutable or mutable depending on the context. In most implementations, it will be presented with a language-appropriate interface, eg, Python `Mapping` or `MutableMapping`.
+
+The following operations manipulate the mapping, providing a getter/setter interface plus the ability to iterate on the collection:
+
+```
+  mapping.get(string key) -> value
+  mapping.has(string key) -> bool
+  mapping.set(string key, simple value) -> void
+  mapping.del(string key) -> void
+  mapping.__iterator__
+  mapping.__length__
+```
+
+> ⚠️ Issue - do we need a "is_readonly" or "is_immutable" attribute, or assume user can figure it out from the context?
+
+> ℹ️ **Note** - it is possible that the data model will grow to include more complex value types. If possible, retain that future option in any API defined.
+
+## SOMACollection
+
+In the following, `soco` refers to an SOMACollection instance or handle.
+
+```
+SOMACollection.create(uri) ->
+SOMACollection.open(uri, mode=read|write) -> soco
+SOMACollection.delete(uri) -> void
+SOMACollection.exists(uri) -> bool   # exists and is a SOMACollection
+soco.close()
+soco.uns -> metadata as a SOMAMapping
+```
+
+> ⚠️ **Issue** - do we need a specific open/close API, or is the object stateless? Or is that language/environment-specific?
```

Comment:
i prefer that open/close be implementation details within the storage layer, so

* API level delete
* Storage level:
  * storage.open
  * storage.delete
  * storage.close
 
since SOMA is an API, not a storage spec, i think notions of open/close don't belong in the API -- open/close is a storage-level issue

Reply:
This idea of delegating (plus/minus some details) is interesting.  Couple of questions:
* what about `exists()` and related?
* what about `create()`?
* I'm not sure I understand `api.delete` above.  Do you mean `update` of existing objects (as in SOMACollection  add/del and SOMADataFrame read/write)?

I think you are suggesting that the SOMACollection, SOMADataFrame and SOMAMatrix lifecycle (create, delete, open, close) are entirely outside the scope of the SOMA API.

Reply:
Apologies!  I misunderstood the original comment. We are aligned on intent - I will improve the prose.